xdsl/dialects/builtin.py

```diff
@@ -395,6 +396,23 @@ def verify_value(self, value: int):
                 f"values in the range [{min_value}, {max_value})"
             )
 
+    def normalize_value(self, value: int) -> int:
+        """
+        Normalize signless values within the expected range to the signed range for the
+        same bitwidth.
+        Values outside the valid range are left as-is.
+        """
+        if self.signedness.data != Signedness.SIGNLESS:
+            return value
+
+        signed_ub = signed_upper_bound(self.bitwidth)
+        unsigned_ub = signed_ub << 1
```

Comment:
```suggestion
        unsigned_ub = unsigned_upper_bound(self.bitwidth)
```
Why not use this? There will be complications with `i0` right?

Reply:
not really, this was a micro-optimisation, but I guess not really very important